internal/model/model.go

```diff
@@ -21,12 +21,18 @@ import (
 	policyv1 "k8s.io/api/policy/v1"
 )
 
+type TaskManagerDesiredState struct {
+	DeploymentType string
+	Deployment     *appsv1.Deployment
+	StatefulSet    *appsv1.StatefulSet
+}
+
 // DesiredClusterState holds desired state of a cluster.
 type DesiredClusterState struct {
 	JmStatefulSet       *appsv1.StatefulSet
 	JmService           *corev1.Service
 	JmIngress           *networkingv1.Ingress
-	TmStatefulSet       *appsv1.StatefulSet
+	TmDesiredState      *TaskManagerDesiredState
```

Comment:
Since there are some changes planned around the observer/reconciler here I would keep things simple and just a `TmDeployment` instead of introducing this new type; No need to add an additional `DeploymentType` to the state since `TmDeployment` and `TmStatefulSet` will be mutually exclusive.

With this, we keep basically all the logic around StatefulSet (observer/reconciler) and just add the necessary bits for the Deployment.

If I'm seeing things right:
```
getDesiredClusterState(observed *ObservedClusterState) *model.DesiredClusterState
```

This should be the only place where we check `TaskManager.DeploymentType` and set either one.


This should simplify things a little bit and will make things easier for what comes next.

Reply:
That would be cleaner indeed. Will do 👍